prover-scheduler/src/scheduler_server.rs

```diff
@@ -0,0 +1,329 @@
+                    if let Ok(scheduler_msg) = self
+                        .handle_gen_batch_proof_response(r.prover_id, scheduler_sender)
+                        .await
+                    {
+                        Ok(scheduler_msg)
+                    } else {
+                        // close the connection
+                        Err(anyhow!("Failed to handle GenBatchProofResponse"))
+                    }
+                }
+                AddServiceResult::Fail(service_id) => {
+                    // close the connection
+                    Err(anyhow!("Failed to add service: {}", service_id))
+                }
+            }
+        } else {
+            // channel closed
+            Err(anyhow!(
+                "Failed to receive AddServiceResult, channel closed"
+            ))
+        }
+    }
+
+    async fn handle_gen_batch_proof_response(
+        &self,
+        provider_id: String,
+        scheduler_sender: mpsc::Sender<Event>,
+    ) -> Result<SchedulerMessage> {
+        let (relay_to, mut relay) = mpsc::channel::<TakeTaskResult>(1);
+        // then, send Event::TriggerTask to the scheduler
+        let event = Event::TakeTask {
+            service_id: provider_id.clone(),
+            relay_to,
+        };
+
+        if let Err(e) = scheduler_sender.send(event.clone()).await {
+            // can't send event to scheduler, close the connection
+            log::error!("Failed to send Event: {:?}, receiver dropped: {}", event, e);
+            return Err(anyhow!(
+                "Failed to send Event: {:?}, receiver dropped: {}",
+                event,
+                e
+            ));
+        }
+
+        // wait for the event result
+        if let Some(take_task_result) = relay.recv().await {
+            match take_task_result {
+                TakeTaskResult::Success(batch_ctx) => {
+                    Ok(SchedulerMessage {
+                        // TODO: received id
+                        id: "".into(),
+                        message_type: Some(
+                            // TODO: impl into trait for BatchContext?
+                            scheduler_message::MessageType::TakeBatchProofTaskResponse(
+                                TakeBatchProofTaskResponse {
+                                    prover_id: provider_id.clone(),
+                                    batch_context_bytes: Some(BatchContextBytes {
+                                        data: serde_json::to_vec(&batch_ctx).unwrap(),
+                                    }),
+                                },
+                            ),
+                        ),
+                    })
+                }
+                TakeTaskResult::Fail(service_id) => {
+                    Err(anyhow!("Failed to take task for service: {}", service_id))
+                }
+            }
+        } else {
+            // channel closed
+            Err(anyhow!("Failed to receive TakeTaskResult, channel closed"))
```

Comment:
Many places are using this expression, change it all to `bail!`.

Reply:
ok

Reply:
done!